st2reactor/st2reactor/rules/filter.py

```diff
@@ -149,7 +157,38 @@ def _render_criteria_pattern(self, criteria_pattern):
             # makes no sense
             return criteria_pattern
 
-        criteria_pattern = render_template_with_system_context(value=criteria_pattern)
+        LOG.debug(
+            'Rendering criteria pattern (%s) with context: %s',
+            criteria_pattern,
+            criteria_context
+        )
+
+        to_complex = False
+
+        # Check if jinja variable is in criteria_pattern and if so lets ensure
+        # the proper type is applied to it using to_complex jinja filter
+        if len(re.findall(r'^{{\s*[A-z0-9_.-]+\s*}}$', criteria_pattern)) > 0:
```

Comment:
It also seems like we use the same regular expression multiple times so it would make sense to compile it once and re-use compiled version (top-level on module load) to speed things up a bit.

On a related note - any ideas if this adds any significant negative performance overhead (I believe I asked a similar thing in other PR, but don't remember the answer)?